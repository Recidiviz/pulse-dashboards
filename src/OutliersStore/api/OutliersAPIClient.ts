// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2023 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { callNewMetricsApi } from "../../api/metrics/metricsClient";
import { MetricBenchmark } from "../models/MetricBenchmark";
import { outliersConfigSchema } from "../models/OutliersConfig";
import { SupervisionOfficer } from "../models/SupervisionOfficer";
import { SupervisionOfficerMetricEvent } from "../models/SupervisionOfficerMetricEvent";
import {
  SupervisionOfficerSupervisor,
  supervisionOfficerSupervisorSchema,
} from "../models/SupervisionOfficerSupervisor";
import type { OutliersStore } from "../OutliersStore";
import { OutliersAPI } from "./interface";

export class OutliersAPIClient implements OutliersAPI {
  // eslint-disable-next-line no-useless-constructor
  constructor(public readonly outliersStore: OutliersStore) {}

  tenantId(): string {
    const { currentTenantId } = this.outliersStore.rootStore;
    if (!currentTenantId) {
      throw new Error(`Attempted to fetch data with undefined tenantId`);
    }
    return currentTenantId;
  }

  init() {
    const endpoint = `outliers/${this.tenantId()}/configuration`.toLowerCase();
    return callNewMetricsApi(
      endpoint,
      this.outliersStore.rootStore.getTokenSilently
    ).then((fetchedData) => {
      return outliersConfigSchema.parse(fetchedData.config);
    });
  }

  /* eslint-disable class-methods-use-this */
  async metricBenchmarks(): Promise<MetricBenchmark[]> {
    const { metricBenchmarksFixture } = await import(
      "../models/offlineFixtures/MetricBenchmarkFixture"
    );
    return metricBenchmarksFixture;
  }

  supervisionOfficerSupervisors(): Promise<SupervisionOfficerSupervisor[]> {
    const endpoint = `outliers/${this.tenantId()}/supervisors`.toLowerCase();
    return callNewMetricsApi(
      endpoint,
      this.outliersStore.rootStore.getTokenSilently
    ).then((fetchedData) => {
      const supervisorData = fetchedData.supervisors as Array<unknown>;
      return supervisorData.map((b) =>
        supervisionOfficerSupervisorSchema.parse(b)
      );
    });
  }

  /* eslint-disable class-methods-use-this */
  async officersForSupervisor(
    supervisorId: string
  ): Promise<Array<SupervisionOfficer>> {
    const { supervisionOfficerFixture } = await import(
      "../models/offlineFixtures/SupervisionOfficerFixture"
    );

    return supervisionOfficerFixture.filter(
      (o) => o.supervisorId === supervisorId
    );
  }

  async supervisionOfficer(officerId: string): Promise<SupervisionOfficer> {
    const { supervisionOfficerFixture } = await import(
      "../models/offlineFixtures/SupervisionOfficerFixture"
    );

    const officerFixture = supervisionOfficerFixture.find(
      (o) => o.externalId === officerId
    );

    if (!officerFixture)
      throw new Error(`Officer ${officerId} not present in fixture data`);

    return officerFixture;
  }

  async supervisionOfficerMetricEvents(
    officerId: string,
    metricId: string
  ): Promise<SupervisionOfficerMetricEvent[]> {
    const { supervisionOfficerMetricEventFixture } = await import(
      "../models/offlineFixtures/SupervisionOfficerMetricEventFixture"
    );
    return supervisionOfficerMetricEventFixture.filter(
      (e) => e.metricId === metricId
    );
  }
}
