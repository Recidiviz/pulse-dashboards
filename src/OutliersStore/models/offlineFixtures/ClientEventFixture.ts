// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2023 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { ClientEvent, clientEventSchema, RawClientEvent } from "../ClientEvent";

export const rawClientEventFixture: Array<RawClientEvent> = [
  {
    metricId: "violations",
    eventDate: "2023-07-22",
    attributes: {
      code: "149",
      description: "NOT ASSOCIATE WITH ANYONE YOU KNOW TO HAVE A FELONY RECORD",
    },
  },
  {
    metricId: "violation_responses",
    eventDate: "2023-05-12",
    attributes: {
      description: "PAROLE REINSTATED",
    },
  },
  {
    metricId: "violations",
    eventDate: "2023-05-01",
    attributes: {
      code: "148",
      description:
        "COMPLY WITH ALCOHOL AND DRUG TESTING ORDERED BY FIELD AGENT",
    },
  },
];

export const clientEventFixture: Array<ClientEvent> = rawClientEventFixture.map(
  (e) => clientEventSchema.parse(e)
);
