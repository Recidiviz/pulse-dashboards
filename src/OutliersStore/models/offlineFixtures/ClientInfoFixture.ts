// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2023 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { mapValues } from "lodash";

import { ClientInfo, clientInfoSchema, RawClientInfo } from "../ClientInfo";

export const rawClientInfoFixture: Record<string, RawClientInfo> = {
  "hashed-707222": {
    clientName: {
      givenNames: "Gary",
      surname: "Alexander",
    },
    clientId: "707222",
    pseudonymizedClientId: "hashed-707222",
    gender: "MALE",
    raceOrEthnicity: "BLACK",
    birthdate: "1971-03-15",
  },
  "hashed-205752": {
    clientName: {
      givenNames: "Beau",
      surname: "Riley",
    },
    clientId: "205752",
    pseudonymizedClientId: "hashed-205752",
    gender: "MALE",
    raceOrEthnicity: "BLACK",
    birthdate: "1971-03-15",
  },
  "hashed-792381": {
    clientName: {
      givenNames: "Deena",
      surname: "Dunlap",
    },
    clientId: "792381",
    pseudonymizedClientId: "hashed-792381",
    gender: "FEMALE",
    raceOrEthnicity: "BLACK",
    birthdate: "1971-03-15",
  },
  "hashed-869516": {
    clientName: {
      givenNames: "Angelo",
      surname: "Cohen",
    },
    clientId: "869516",
    pseudonymizedClientId: "hashed-869516",
    gender: "MALE",
    raceOrEthnicity: "BLACK",
    birthdate: "1971-03-15",
  },
  "hashed-170571": {
    clientName: {
      givenNames: "Ward",
      surname: "Bradley",
    },
    clientId: "170571",
    pseudonymizedClientId: "hashed-170571",
    gender: "MALE",
    raceOrEthnicity: "BLACK",
    birthdate: "1971-03-15",
  },
  "hashed-461718": {
    clientName: {
      givenNames: "Janette",
      surname: "Sosa",
    },
    clientId: "461718",
    pseudonymizedClientId: "hashed-461718",
    gender: "FEMALE",
    raceOrEthnicity: "BLACK",
    birthdate: "1971-03-15",
  },
  "hashed-837771": {
    clientName: {
      givenNames: "Nora",
      surname: "Robbins",
    },
    clientId: "837771",
    pseudonymizedClientId: "hashed-837771",
    gender: null,
    raceOrEthnicity: null,
    birthdate: null,
  },
  "hashed-617754": {
    clientName: {
      givenNames: "Jason",
      surname: "Barton",
    },
    clientId: "617754",
    pseudonymizedClientId: "hashed-617754",
    gender: "MALE",
    raceOrEthnicity: "BLACK",
    birthdate: "1971-03-15",
  },

  "hashed-985771": {
    clientName: {
      givenNames: "Nicholas",
      surname: "Rhodes",
    },
    clientId: "985771",
    pseudonymizedClientId: "hashed-985771",
    gender: "MALE",
    raceOrEthnicity: "BLACK",
    birthdate: "1971-03-15",
  },
  "hashed-326160": {
    clientName: {
      givenNames: "Ethel",
      surname: "Leonard",
    },
    clientId: "326160",
    pseudonymizedClientId: "hashed-326160",
    gender: "FEMALE",
    raceOrEthnicity: "BLACK",
    birthdate: "1971-03-15",
  },
  "hashed-931890": {
    clientName: {
      givenNames: "Ricardo",
      surname: "Wood",
    },
    clientId: "931890",
    pseudonymizedClientId: "hashed-931890",
    gender: "MALE",
    raceOrEthnicity: "BLACK",
    birthdate: "1971-03-15",
  },
  "hashed-930441": {
    clientName: {
      givenNames: "Stanley",
      surname: "Maxwell",
    },
    clientId: "930441",
    pseudonymizedClientId: "hashed-930441",
    gender: "MALE",
    raceOrEthnicity: "BLACK",
    birthdate: "1971-03-15",
  },
  "hashed-564940": {
    clientName: {
      givenNames: "Miguel",
      surname: "Haynes",
    },
    clientId: "564940",
    pseudonymizedClientId: "hashed-564940",
    gender: "MALE",
    raceOrEthnicity: "BLACK",
    birthdate: "1971-03-15",
  },

  "hashed-144925": {
    clientName: {
      givenNames: "Inez",
      surname: "Griffith",
    },
    clientId: "144925",
    pseudonymizedClientId: "hashed-144925",
    gender: "FEMALE",
    raceOrEthnicity: "BLACK",
    birthdate: "1971-03-15",
  },
  "hashed-283496": {
    clientName: {
      givenNames: "Gerald",
      surname: "Barber",
    },
    clientId: "283496",
    pseudonymizedClientId: "hashed-283496",
    gender: "MALE",
    raceOrEthnicity: "BLACK",
    birthdate: "1971-03-15",
  },
  "hashed-334531": {
    clientName: {
      givenNames: "Leroy",
      surname: "Stone",
    },
    clientId: "334531",
    pseudonymizedClientId: "hashed-334531",
    gender: "MALE",
    raceOrEthnicity: "BLACK",
    birthdate: "1971-03-15",
  },
  "hashed-840134": {
    clientName: {
      givenNames: "Patrick",
      surname: "Schneider",
    },
    clientId: "840134",
    pseudonymizedClientId: "hashed-840134",
    gender: "MALE",
    raceOrEthnicity: "BLACK",
    birthdate: "1971-03-15",
  },
  "hashed-968048": {
    clientName: {
      givenNames: "Violet",
      surname: "Johnson",
    },
    clientId: "968048",
    pseudonymizedClientId: "hashed-968048",
    gender: "FEMALE",
    raceOrEthnicity: "BLACK",
    birthdate: "1971-03-15",
  },
  "hashed-413899": {
    clientName: {
      givenNames: "Warren",
      surname: "Franklin",
    },
    clientId: "413899",
    pseudonymizedClientId: "hashed-413899",
    gender: "MALE",
    raceOrEthnicity: "BLACK",
    birthdate: "1971-03-15",
  },
  "hashed-415630": {
    clientName: {
      givenNames: "Alta",
      surname: "Jackson",
    },
    clientId: "415630",
    pseudonymizedClientId: "hashed-415630",
    gender: "FEMALE",
    raceOrEthnicity: "BLACK",
    birthdate: "1971-03-15",
  },

  "hashed-235261": {
    clientName: {
      givenNames: "Rosie",
      surname: "Luna",
    },
    clientId: "235261",
    pseudonymizedClientId: "hashed-235261",
    gender: "FEMALE",
    raceOrEthnicity: "NATIVE_HAWAIIAN_PACIFIC_ISLANDER",
    birthdate: "1971-03-15",
  },
  "hashed-531407": {
    clientName: {
      givenNames: "Glenn",
      surname: "Francis",
    },
    clientId: "531407",
    pseudonymizedClientId: "hashed-531407",
    gender: "MALE",
    raceOrEthnicity: "BLACK",
    birthdate: "1971-03-15",
  },
  "hashed-436502": {
    clientName: {
      givenNames: "Steve",
      surname: "Schwartz",
    },
    clientId: "436502",
    pseudonymizedClientId: "hashed-436502",
    gender: "MALE",
    raceOrEthnicity: "BLACK",
    birthdate: "1971-03-15",
  },
  "hashed-339710": {
    clientName: {
      givenNames: "Bertha",
      surname: "Figueroa",
    },
    clientId: "339710",
    pseudonymizedClientId: "hashed-339710",
    gender: "FEMALE",
    raceOrEthnicity: "AMERICAN_INDIAN_ALASKAN_NATIVE",
    birthdate: "1971-03-15",
  },

  "hashed-803013": {
    clientName: {
      givenNames: "Troy",
      surname: "Rodgers",
    },
    clientId: "803013",
    pseudonymizedClientId: "hashed-803013",
    gender: "MALE",
    raceOrEthnicity: "BLACK",
    birthdate: "1971-03-15",
  },
  "hashed-470308": {
    clientName: {
      givenNames: "Leo",
      surname: "Marshall",
    },
    clientId: "470308",
    pseudonymizedClientId: "hashed-470308",
    gender: "MALE",
    raceOrEthnicity: "BLACK",
    birthdate: "1971-03-15",
  },
  "hashed-582059": {
    clientName: {
      givenNames: "Shawn",
      surname: "Romero",
    },
    clientId: "582059",
    pseudonymizedClientId: "hashed-582059",
    gender: "MALE",
    raceOrEthnicity: "BLACK",
    birthdate: "1971-03-15",
  },
  "hashed-374673": {
    clientName: {
      givenNames: "Jeremiah",
      surname: "Walsh",
    },
    clientId: "374673",
    pseudonymizedClientId: "hashed-374673",
    gender: "MALE",
    raceOrEthnicity: "BLACK",
    birthdate: "1971-03-15",
  },
  "hashed-128785": {
    clientName: {
      givenNames: "Dorothy",
      surname: "Luna",
    },
    clientId: "128785",
    pseudonymizedClientId: "hashed-128785",
    gender: "FEMALE",
    raceOrEthnicity: "BLACK",
    birthdate: "1971-03-15",
  },
  "hashed-914408": {
    clientName: {
      givenNames: "Viola",
      surname: "Schultz",
    },
    clientId: "914408",
    pseudonymizedClientId: "hashed-914408",
    gender: "MALE",
    raceOrEthnicity: "BLACK",
    birthdate: "1971-03-15",
  },
  "hashed-504894": {
    clientName: {
      givenNames: "Marie",
      surname: "Hudson",
    },
    clientId: "504894",
    pseudonymizedClientId: "hashed-504894",
    gender: "FEMALE",
    raceOrEthnicity: "BLACK",
    birthdate: "1971-03-15",
  },
};

export const clientInfoFixture: Record<string, ClientInfo> = mapValues(
  rawClientInfoFixture,
  (c) => clientInfoSchema.parse(c)
);
