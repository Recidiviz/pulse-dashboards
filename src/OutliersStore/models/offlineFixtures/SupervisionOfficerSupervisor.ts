// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2023 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import {
  RawSupervisionOfficerSupervisor,
  supervisionOfficerSupervisorSchema,
} from "../SupervisionOfficerSupervisor";

export const rawSupervisionOfficerSupervisorFixture: RawSupervisionOfficerSupervisor[] =
  [
    {
      fullName: { givenNames: "Miles", middleNames: "D", surname: "Davis" },
      externalId: "mdavis123",
      supervisionDistrict: "Region D1",
      pseudonymizedId: "hashed-mdavis123",
      hasOutliers: true,
    },
    {
      fullName: { givenNames: "Billie", surname: "Holiday" },
      externalId: "bholiday456",
      supervisionDistrict: null,
      pseudonymizedId: "hashed-bholiday456",
      hasOutliers: true,
    },
    {
      fullName: { givenNames: "Ella", surname: "Fitzgerald" },
      externalId: "efitzgerald789",
      supervisionDistrict: "D1",
      pseudonymizedId: "hashed-efitzgerald789",
      hasOutliers: false,
    },
    {
      fullName: { givenNames: "John", surname: "Coltrane" },
      externalId: "jColtrane321",
      supervisionDistrict: null,
      pseudonymizedId: "hashed-jColtrane321",
      hasOutliers: false,
    },
  ];
export const supervisionOfficerSupervisorsFixture =
  rawSupervisionOfficerSupervisorFixture.map((b) =>
    supervisionOfficerSupervisorSchema.parse(b)
  );
