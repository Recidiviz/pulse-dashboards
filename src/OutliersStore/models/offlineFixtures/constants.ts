// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2023 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { ascending, range } from "d3-array";
import { subMonths } from "date-fns";
import { z } from "zod";

export const ADVERSE_METRIC_IDS = z.enum([
  "incarceration_starts",
  "absconsions_bench_warrants",
  "incarceration_starts_technical_violation",
]);

export const CASELOAD_TYPE_IDS = z.enum(["GENERAL_OR_OTHER", "SEX_OFFENSE"]);

export const LATEST_END_DATE = new Date(2023, 8, 1);

export const LOOKBACK_END_DATES = range(6)
  .map((offset) => subMonths(LATEST_END_DATE, offset))
  .sort(ascending);
