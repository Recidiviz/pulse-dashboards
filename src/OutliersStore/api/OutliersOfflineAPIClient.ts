/* eslint-disable class-methods-use-this */
// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2023 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { differenceInDays, subDays } from "date-fns";

import { ClientEvent } from "../models/ClientEvent";
import { ClientInfo } from "../models/ClientInfo";
import { MetricBenchmark } from "../models/MetricBenchmark";
import { LATEST_END_DATE } from "../models/offlineFixtures/constants";
import { SupervisionOfficer } from "../models/SupervisionOfficer";
import { SupervisionOfficerMetricEvent } from "../models/SupervisionOfficerMetricEvent";
import { SupervisionOfficerSupervisor } from "../models/SupervisionOfficerSupervisor";
import type { OutliersStore } from "../OutliersStore";
import { OutliersAPI } from "./interface";

export class OutliersOfflineAPIClient implements OutliersAPI {
  // eslint-disable-next-line no-useless-constructor
  constructor(public readonly outliersStore: OutliersStore) {}

  async init() {
    const config = (
      await import("../models/offlineFixtures/OutliersConfigFixture")
    ).OutliersConfigFixture;

    return config;
  }

  async metricBenchmarks(): Promise<MetricBenchmark[]> {
    const { metricBenchmarksFixture } = await import(
      "../models/offlineFixtures/MetricBenchmarkFixture"
    );
    return metricBenchmarksFixture;
  }

  async supervisionOfficerSupervisors(): Promise<
    SupervisionOfficerSupervisor[]
  > {
    const { supervisionOfficerSupervisorsFixture } = await import(
      "../models/offlineFixtures/SupervisionOfficerSupervisor"
    );
    return supervisionOfficerSupervisorsFixture;
  }

  async officersForSupervisor(
    supervisorPseudoId: string
  ): Promise<Array<SupervisionOfficer>> {
    const { supervisionOfficerFixture } = await import(
      "../models/offlineFixtures/SupervisionOfficerFixture"
    );

    return supervisionOfficerFixture.filter(
      (o) => `hashed-${o.supervisorExternalId}` === supervisorPseudoId
    );
  }

  async supervisionOfficer(
    officerPseudoId: string
  ): Promise<SupervisionOfficer> {
    const { supervisionOfficerFixture } = await import(
      "../models/offlineFixtures/SupervisionOfficerFixture"
    );

    const officerFixture = supervisionOfficerFixture.find(
      (o) => o.pseudonymizedId === officerPseudoId
    );

    if (!officerFixture)
      throw new Error(`Officer ${officerPseudoId} not present in fixture data`);

    return officerFixture;
  }

  async supervisionOfficerMetricEvents(
    officerPseudoId: string,
    metricId: string
  ): Promise<SupervisionOfficerMetricEvent[]> {
    const { supervisionOfficerMetricEventFixture } = await import(
      "../models/offlineFixtures/SupervisionOfficerMetricEventFixture"
    );
    return supervisionOfficerMetricEventFixture.filter(
      (e) => e.metricId === metricId
    );
  }

  async clientInfo(clientPseudoId: string): Promise<ClientInfo> {
    const { clientInfoFixture } = await import(
      "../models/offlineFixtures/ClientInfoFixture"
    );

    const clientInfo = clientInfoFixture[clientPseudoId];

    if (!clientInfo) {
      throw new Error(`Client ${clientPseudoId} not present in fixture data`);
    }

    return clientInfo;
  }

  async clientEvents(
    clientPseudoId: string,
    endDate: Date
  ): Promise<Array<ClientEvent>> {
    const { clientEventFixture } = await import(
      "../models/offlineFixtures/ClientEventFixture"
    );

    // adjust fixture dates relative to the input
    const dateOffsetInDays = differenceInDays(LATEST_END_DATE, endDate);

    return clientEventFixture.map((event) => ({
      ...event,
      eventDate: subDays(event.eventDate, dateOffsetInDays),
    }));
  }
}
