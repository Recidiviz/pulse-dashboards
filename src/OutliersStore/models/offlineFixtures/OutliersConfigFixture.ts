// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2023 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { OutliersConfig } from "../OutliersConfig";
import { ADVERSE_METRIC_IDS } from "./constants";

export const OutliersConfigFixture: OutliersConfig = {
  supervisionOfficerLabel: "officer",
  supervisionDistrictLabel: "region",
  supervisionDistrictManagerLabel: "district manager",
  supervisionJiiLabel: "client",
  supervisionSupervisorLabel: "supervisor",
  supervisionUnitLabel: "team",
  learnMoreUrl: "https://recidiviz.org",
  metrics: [
    {
      name: ADVERSE_METRIC_IDS.enum.absconsions_bench_warrants,
      bodyDisplayName: "absconsion rate",
      titleDisplayName: "Absconsion Rate",
      eventName: "absconsions",
      outcomeType: "ADVERSE",
    },
    {
      name: ADVERSE_METRIC_IDS.enum.incarceration_starts,
      bodyDisplayName: "incarceration rate",
      titleDisplayName: "Incarceration Rate",
      eventName: "incarcerations",
      outcomeType: "ADVERSE",
    },
    {
      name: ADVERSE_METRIC_IDS.enum.incarceration_starts_technical_violation,
      bodyDisplayName: "technical incarceration rate",
      titleDisplayName: "Technical Incarceration Rate",
      eventName: "technical incarcerations",
      outcomeType: "ADVERSE",
    },
  ],
  clientEvents: [
    {
      displayName: "Violations",
      name: "violations",
    },
    {
      displayName: "Sanctions",
      name: "violation_responses",
    },
  ],
};
