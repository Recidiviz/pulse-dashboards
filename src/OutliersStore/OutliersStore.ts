// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2023 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { makeAutoObservable, reaction, runInAction } from "mobx";

import { RootStore } from "../RootStore";
import { isDemoMode } from "../utils/isDemoMode";
import { isOfflineMode } from "../utils/isOfflineMode";
import { isTestEnv } from "../utils/isTestEnv";
import { OutliersAPI } from "./api/interface";
import { OutliersAPIClient } from "./api/OutliersAPIClient";
import { OutliersOfflineAPIClient } from "./api/OutliersOfflineAPIClient";
import { OutliersSupervisionStore } from "./stores/OutliersSupervisionStore";
import { FlowMethod } from "./types";

export class OutliersStore {
  supervisionStore?: OutliersSupervisionStore;

  apiClient: OutliersAPI;

  constructor(public rootStore: RootStore) {
    this.apiClient =
      isOfflineMode() || isTestEnv() || isDemoMode()
        ? new OutliersOfflineAPIClient(this)
        : new OutliersAPIClient(this);

    makeAutoObservable(this);

    // reset the store for each new tenant
    reaction(
      () => this.rootStore.currentTenantId,
      () => {
        runInAction(() => this.reset());
      }
    );
  }

  reset() {
    this.supervisionStore = undefined;
  }

  /**
   * Creates this.supervisionStore for the current tenant, if it does not already exist.
   * This is a MobX flow method and should be called with mobx.flowResult.
   */
  *hydrateSupervisionStore(): FlowMethod<OutliersAPI["init"], void> {
    if (this.supervisionStore) return;

    const config = yield this.apiClient.init();

    this.supervisionStore = new OutliersSupervisionStore(this, config);
  }
}
