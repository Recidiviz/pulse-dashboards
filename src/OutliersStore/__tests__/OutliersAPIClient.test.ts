// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2023 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { disableFetchMocks, enableFetchMocks } from "jest-fetch-mock";

import { RootStore } from "../../RootStore";
import { OutliersAPIClient } from "../api/OutliersAPIClient";
import { OutliersConfigFixture } from "../models/offlineFixtures/OutliersConfigFixture";
import { supervisionOfficerSupervisorsFixture } from "../models/offlineFixtures/SupervisionOfficerSupervisor";
import { OutliersStore } from "../OutliersStore";

const mockTenantId = "us_tn";
const BASE_URL = `http://localhost:5000/outliers/${mockTenantId}`;
const OLD_ENV = process.env;

describe("OutliersAPIClient", () => {
  let client: OutliersAPIClient;

  beforeAll(() => {
    enableFetchMocks();
    process.env = Object.assign(process.env, {
      REACT_APP_DEPLOY_ENV: "dev",
      REACT_APP_NEW_BACKEND_API_URL: "http://localhost:5000",
    });
    // @ts-ignore
    const mockRootStore = {
      getTokenSilently: jest.fn(),
      currentTenantId: mockTenantId,
    } as RootStore;
    client = new OutliersAPIClient(new OutliersStore(mockRootStore));
  });

  afterEach(() => {
    fetchMock.resetMocks();
  });

  afterAll(() => {
    process.env = OLD_ENV;
    disableFetchMocks();
  });

  it("init calls the correct endpoint", async () => {
    fetchMock.mockResponse(JSON.stringify({ config: OutliersConfigFixture }));
    await client.init();
    expect(fetchMock.mock.calls[0][0]).toEqual(
      encodeURI(`${BASE_URL}/configuration`)
    );
  });

  it("init parses the config data", async () => {
    fetchMock.mockResponse(JSON.stringify({ config: OutliersConfigFixture }));
    const response = await client.init();
    expect(response).toMatchSnapshot();
  });

  it("supervisionOfficerSupervisors calls the correct endpoint", async () => {
    fetchMock.mockResponse(JSON.stringify({ supervisors: [] }));
    await client.supervisionOfficerSupervisors();
    expect(fetchMock.mock.calls[0][0]).toEqual(
      encodeURI(`${BASE_URL}/supervisors`)
    );
  });

  it("supervisionOfficerSupervisors parses the data", async () => {
    fetchMock.mockResponse(
      JSON.stringify({ supervisors: supervisionOfficerSupervisorsFixture })
    );
    const response = await client.supervisionOfficerSupervisors();
    expect(response).toMatchSnapshot();
  });

  it("supervisionOfficerSupervisors throws error if tenantId is undefined", async () => {
    // @ts-ignore
    const mockRootStore = {
      getTokenSilently: jest.fn(),
      currentTenantId: undefined,
    } as RootStore;
    const badClient = new OutliersAPIClient(new OutliersStore(mockRootStore));
    expect(() => badClient.supervisionOfficerSupervisors()).toThrow(
      "Attempted to fetch data with undefined tenantId"
    );
  });
});
