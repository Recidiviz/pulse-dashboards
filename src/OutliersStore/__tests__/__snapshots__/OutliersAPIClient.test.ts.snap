// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OutliersAPIClient init parses the config data 1`] = `
Object {
  "learnMoreUrl": "https://recidiviz.org",
  "metrics": Array [
    Object {
      "bodyDisplayName": "absconsion rate",
      "eventName": "absconsions",
      "name": "absconsions_bench_warrants",
      "outcomeType": "ADVERSE",
      "titleDisplayName": "Absconsion Rate",
    },
    Object {
      "bodyDisplayName": "incarceration rate",
      "eventName": "incarcerations",
      "name": "incarceration_starts",
      "outcomeType": "ADVERSE",
      "titleDisplayName": "Incarceration Rate",
    },
    Object {
      "bodyDisplayName": "technical incarceration rate",
      "eventName": "technical incarcerations",
      "name": "incarceration_starts_technical_violation",
      "outcomeType": "ADVERSE",
      "titleDisplayName": "Technical Incarceration Rate",
    },
  ],
  "supervisionOfficerLabel": "officer",
}
`;

exports[`OutliersAPIClient supervisionOfficerSupervisors parses the data 1`] = `
Array [
  Object {
    "displayName": "Miles D Davis",
    "externalId": "mdavis123",
    "fullName": Object {
      "givenNames": "Miles",
      "middleNames": "D",
      "surname": "Davis",
    },
    "hasOutliers": true,
    "pseudonymizedId": "hashed-mdavis123",
    "supervisionDistrict": "D1",
  },
  Object {
    "displayName": "Billie Holiday",
    "externalId": "bholiday456",
    "fullName": Object {
      "givenNames": "Billie",
      "surname": "Holiday",
    },
    "hasOutliers": true,
    "pseudonymizedId": "hashed-bholiday456",
    "supervisionDistrict": null,
  },
  Object {
    "displayName": "Ella Fitzgerald",
    "externalId": "efitzgerald789",
    "fullName": Object {
      "givenNames": "Ella",
      "surname": "Fitzgerald",
    },
    "hasOutliers": false,
    "pseudonymizedId": "hashed-efitzgerald789",
    "supervisionDistrict": "D1",
  },
  Object {
    "displayName": "John Coltrane",
    "externalId": "jColtrane321",
    "fullName": Object {
      "givenNames": "John",
      "surname": "Coltrane",
    },
    "hasOutliers": false,
    "pseudonymizedId": "hashed-jColtrane321",
    "supervisionDistrict": null,
  },
]
`;
